src/GraphCoarsenBuilder.cpp

```diff
@@ -190,17 +210,255 @@ void AssembleMBuilder::AddTraceAcross(int row, int col, double value)
     CoarseM_->Add(row, col, value);
 }
 
-std::unique_ptr<mfem::SparseMatrix> ElementMBuilder::GetCoarseM()
+std::unique_ptr<mfem::SparseMatrix> ElementMBuilder::GetCoarseM(
+    const mfem::Vector& fineMdiag,
+    const mfem::SparseMatrix& Pedges, const mfem::SparseMatrix& face_cdof)
 {
     return std::unique_ptr<mfem::SparseMatrix>(nullptr);
 }
 
-std::unique_ptr<mfem::SparseMatrix> AssembleMBuilder::GetCoarseM()
+std::unique_ptr<mfem::SparseMatrix> AssembleMBuilder::GetCoarseM(
+    const mfem::Vector& fineMdiag,
+    const mfem::SparseMatrix& Pedges, const mfem::SparseMatrix& face_cdof)
 {
     CoarseM_->Finalize(0);
     return std::move(CoarseM_);
 }
 
+void CoefficientMBuilder::SetCoefficient(const mfem::Vector& agg_weight_inverse)
+{
+    agg_weights_.SetSize(agg_weight_inverse.Size());
+    for (int i = 0; i < agg_weights_.Size(); ++i)
+    {
+        agg_weights_[i] = 1.0 / agg_weight_inverse[i];
+    }
+}
+
+/// this method may be unnecessary, could just use GetTableRow()
+void CoefficientMBuilder::GetCoarseFaceDofs(
+    const mfem::SparseMatrix& face_cdof, int face, mfem::Array<int>& local_coarse_dofs) const
+{
+    mfem::Array<int> temp;
+    GetTableRow(face_cdof, face, temp); // returns a writeable reference
+    temp.Copy(local_coarse_dofs); // make sure we do not modify the matrix
+}
+
+void CoefficientMBuilder::GetCoarseAggDofs(int agg, mfem::Array<int>& local_coarse_dofs) const
+{
+    int agg_size = coarse_agg_dof_offsets_[agg + 1] - coarse_agg_dof_offsets_[agg];
+    local_coarse_dofs.SetSize(agg_size);
+    for (int i = 0; i < agg_size; ++i)
+    {
+        local_coarse_dofs[i] = coarse_agg_dof_offsets_[agg] + i;
+    }
+}
+
+mfem::DenseMatrix CoefficientMBuilder::RTDP(const mfem::DenseMatrix& R,
+                                            const mfem::Vector& D,
+                                            const mfem::DenseMatrix& P)
+{
+    mfem::DenseMatrix out(R.Width(), P.Width());
+    mfem::DenseMatrix Rt;
+    Rt.Transpose(R);
+    Rt.RightScaling(D);
+    Mult(Rt, P, out);
```

Comment:
MFEM w/ LAPACK will break if these have a dimension of size zero.  We can check for this and return early.